### PR TITLE
TEP-0114: Tests Wait Custom Task without CRD in PipelineRun

### PR DESCRIPTION
**Pull request by XinruZhang**
# Changes

This commit starts to test the behavior of Custom Task Run **without
CRD** in a **PipelineRun** against **a Custom Task Controller**.

Future Work: test Custom Task Run **with CRD** against the Wait Custom
Task Controller.

/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

``` release-note
NONE
```

### CODE REVIEW

**Comment by lbernick** [comment]
I'm a bit confused what the goal of these tests are-- what do you mean by with/without CRD? The name of the test class is "TestWaitCustomTask_PipelineRun" but the the test cases are PipelineRun running/succeeded/timed out and it seems like it's testing interaction between pipelinerun timeout and wait task duration. Could you please just give a few sentences about what the behavior being tested is?

**Comment by XinruZhang** [comment]
> I'm a bit confused what the goal of these tests are-- what do you mean by with/without CRD? The name of the test class is "TestWaitCustomTask_PipelineRun" but the the test cases are PipelineRun running/succeeded/timed out and it seems like it's testing interaction between pipelinerun timeout and wait task duration. Could you please just give a few sentences about what the behavior being tested is?

Thanks for the comment Lee @lbernick, great point. I think the main purpose is to test how the run status is reflected upon different PipelineRun statuses. I'll update the testing name to make it clearer.

About with/without CRD, the difference is whether the Custom Task has it's own spec&status defined. We can take the existing  `wait-task` controller as an example, we only need to specify `apiversion`, `kind` and an input param `duration` to make it work. The `wait.testing.tekton.dev/v1alpha1/Wait` doesn't have its own spec and status fields. Therefore the Run status doesn't change upon Wait.status changes.

Shortly speaking, we are basically mimic the actual usage of Custom Task Run.

wdyt? Let me know there's still confusion :)

**Comment by jerop** [comment]
@lbernick @XinruZhang adding a comment to clarify further.

There are some `Custom Tasks` that don't define CRDs - they are only used inline in `Runs` and `PipelineRuns` - these are referred to as "unnamed" `Custom Tasks` in [TEP-0002](https://github.com/tektoncd/community/blob/main/teps/0002-custom-tasks.md). Examples of these are [`Wait`](https://github.com/tektoncd/experimental/tree/f60e1cd8ce22ed745e335f6f547bb9a44580dc7c/wait-task) and [`CEL`](https://github.com/tektoncd/experimental/tree/f60e1cd8ce22ed745e335f6f547bb9a44580dc7c/cel). 

There are some `Custom Tasks` that define CRDs e.g. [`TaskLoop`](https://github.com/tektoncd/experimental/tree/f60e1cd8ce22ed745e335f6f547bb9a44580dc7c/task-loops) and [`PipelineLoop`](https://github.com/tektoncd/experimental/tree/f60e1cd8ce22ed745e335f6f547bb9a44580dc7c/pipeline-loops).

These end-to-end tests cover the first scenario (without CRDs) and this may be sufficient for now. If needed later, we may consider adding another `Custom Task` for testing that covers the second scenario.

File: test/custom_task_test.go [hunk 1]
```diff
@@ -51,10 +51,18 @@ const (
...
-	"enable-custom-tasks": "true",
-	"enable-api-fields":   "alpha",
-}
+var (
+	supportedFeatureGates = map[string]string{
+		"enable-custom-tasks": "true",
+		"enable-api-fields":   "alpha",
+	}
+
+	filterTypeMeta          = cmpopts.IgnoreFields(metav1.TypeMeta{}, "Kind", "APIVersion")
```
Comment (jerop) [inline test/custom_task_test.go:60]: Why do we filter `Kind` and `APIVersion`? 

Makes sense to filter the random fields, such as `CreationTimestamp` and `CompletionTime`, but it's not clear why we filter these fields that are known

Comment (XinruZhang) [reply test/custom_task_test.go:60]: The actual resources retrieved from API server don't specify Kind and APIVersion. Do you think It's better to manually set the two fields in wantResource (like wantPipelineRun or wantRun) as empty strings?

Comment (jerop) [reply test/custom_task_test.go:60]: ah makes sense, checked that other tests also do the same, thanks!

**Review by jerop — APPROVED** [review]
Thanks @XinruZhang!

**Comment by lbernick** [comment]
ok thanks @XinruZhang  @jerop for the explanation, this makes sense.

/lgtm

**Pull request merged**